### ISSUE TITLE
Python 3: Session uid's are built to wrong type

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8215]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| SamyCookie reported|
|-|-|
|Trac ID|trac#8215|
|Type|defect|
|Created|2016-02-19 12:15:24Z|
|Branch|https://github.com/twisted/twisted/tree/http-session-id-type-8215|

As 'web.server.Session' documentation said at line 547, its uid should be 'bytes'.
However in Python 3 they are made to 'str'.
This behaviour causes issues with cookies retrieval (see [#8077](https://github.com/twisted/twisted/issues/8077)).

see my proposed solution at
https://github.com/SamyCookie/twisted/commit/9ca7d578536e47baa6bb0b0e75e5bceb47b5ff65

Attachments:

* [wrong_uid_type-8215-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b19/b196a60e725ce8d8d1d56d04e253b6eed0eeb778/da7527d3ac24c61709af44ae3c114881dbce0157.patch) (1736 bytes) - added by SamyCookie on 2016-02-19 12:24:37Z - 
* [8215.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b19/b196a60e725ce8d8d1d56d04e253b6eed0eeb778/1eae44000f9bc6a32915f0d3571f01db7e7ea4ba.diff) (2226 bytes) - added by SamyCookie on 2016-03-04 12:09:18Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8215 8215
type__defect defect
reporter__SamyCookie SamyCookie
priority__normal normal
milestone__Twisted_16_1 Twisted 16.1
branch__branches_http_session_id_type_8215 branches/http-session-id-type-8215
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1455884124709558 1455884124709558
changetime__1457339839533031 1457339839533031
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|SamyCookie commented|
|-|-|

Any thought about this ticket ?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

This is a bit of a nasty bug that slipped under my radar.

It's not going to block 16.0 but I'll file it under things to block 16.1.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to SamyCookie|
|-|-|

Hi,

Thanks for the patch.

If you want feedback for your patch make sure the ticket is listed in the review queue https://twistedmatrix.com/trac/report/25 ... that is, it has the "review" keyword

Major comments:

1. I think that the new test should be writen against makeSession and getSession as this is the public API and this should be the thing which breaks.

You can test against ```self.sessions[uid]``` ... not sure why this is public... but I would say that self.session should be make private and only use ``` self.getSession ```

2. The patch also needs a news file fragment so that the fix is advertise in the release notes. see https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

Minor comments:

1. I would prefer to see a new test on the public API which demonstrate how the current code fails for your usage :)

2. Do we really need the local imports?

Please review my comments and resubmit a new patch.

Thanks again

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

put back the milestone. sorry.

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|SamyCookie commented|
|-|-|

see uploaded patch or https://github.com/twisted/twisted/compare/trunk...SamyCookie:%238215

Major 1 and 2: done
Minor 1: it should be done in [#8077](https://github.com/twisted/twisted/issues/8077). This issue is only a minor part of this one.
Minor 2: as you want, I just kept the previous written logic.

additionnal note: for more consistency, may it be interesting to use and store UUID type from Python standard library https://docs.python.org/2/library/uuid.html ?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

I assume your latest patch is ready for review.

Please don't forget to add the "review" keyword.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

(In [46915]) Branching to http-session-id-type-8215.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

Thanks for the update.

Minor comments:

1. The news fragment are fragment of our release notes and their tone should be that of a release note and not of a commit message. The wiki page contains a few examples and documentation about how to write those fragments.

2. getSession documentation should be updated to inform that uid is of type bytes.

3. test_getSession is in fact 2 tests :)

4. Please avoid reusing the same variable name for different things and try to use names which describe the purpose / usage of a variable :)

I have update the branch based on the minor comments, so there is no required action on your side.. other than reviewing my changes :)

Since I have also made changes to this branch, it will need to be reviewed by someone else.

Will try to find another reviewer to check that changes. Hope this can be done soon.

Thanks again for your contribution. Much appreciated.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi,

Two things:

- getSession's docstring needs an L{} where "keyerror" is mentioned
- the test docstrings shouldn't use "it" (that is, change "it" to L{site.getSession})

Fix these two then merge. Thanks SamyCookie and Adi for the work on this.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46924]) Merge http-session-id-type-8215: In twisted.web use bytes for session id.

Authors: SamyCookie, adiroiban
Reviewers: adiroiban, hawkowl
Fixes: [#8215](https://github.com/twisted/twisted/issues/8215)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Noticed that the topfile for this change wasn't placed in the correct topfiles folder.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

Gah -- thanks for spotting that. Added an item in the release managers notes (https://twistedmatrix.com/trac/wiki/ReleaseManagerNotes) so I can fix this in the 16.1 release.

**Comment by twisted-trac** [comment]
|<img alt="SamyCookie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|SamyCookie commented|
|-|-|

Oops, sorry for the wrong topfile location, I'll be more careful next time :/
Thanks for helping me and merging this.
